### PR TITLE
implement the engine bits for debugging support

### PR DESCRIPTION
**Pull request by tgummerer**
We want to introduce dubugging support for Pulumi programs.  This PR implements the engine changes necessary for that.  Namely, we pass the information whether we expect a debugger to be started to the language runtime, and introduce a way for the language runtime plugins to report to the engine that we're waiting for a debugger to attach. The language runtime is expected to include the information relevant for the user to be able to attach to the debugger, as well as a shortened message.

The idea is that the configuration can be picked up by an IDE, and the debugger can attach automatically.  Meanwhile the short message should contain enough information to be able to attach a debugger manually, while being short enough to be displayed to the user in the CLI output. (this will commonly be either the port of the debugger, or the PID of the process being debugged).

The implementation of the CLI flags and each of the language runtimes will follow in subsequent PRs.

I tried adding a test to this, but I'm not sure it's possible with our current testing infrastructure.  To do this properly, we'd need to make a RPC call to the engine, but we don't have that available in the lifecycletests (? please let me know if I'm missing something).  Once more of the feature is implemented we might be able to implement an integration test for it.  (Not straightforward either, as we'll have to tell the debugger to continue, but that should be more doable).

Another thing that's not clear to me is that @EronWright mentions this could be used for MLC/provider debugging as well.  However I'm not seeing how that's going to work, as MLCs/providers are being run as a binary plugin, which we don't compile from pulumi/pulumi, and thus wouldn't necessarily even know which debugger to launch it under without a bunch of additional configuration, that might be better in a shim around the program (or just keeping the debugging the way we're currently doing, launching the program and then letting the engine attach to it).

### CODE REVIEW

**Comment by Frassle** [comment]
>  as MLCs/providers are being run as a binary plugin, which we don't compile from pulumi/pulumi, and thus wouldn't necessarily even know which debugger to launch it under without a bunch of additional configuration

Not always. Shimless will let you launch an MLC/provider from source. See the "RunPlugin" method on the language runtime grpc server.

**Comment by tgummerer** [comment]
> Not always. Shimless will let you launch an MLC/provider from source. See the "RunPlugin" method on the language runtime grpc server.

Ah great!  I don't think we want to do that right now anyway, but I wanted to make sure it can still be done with what we are doing here.  Looking at it that's a separate RPC to start the plugin, so we'll need to add a new argument there anyway, and the protocol changes we have right now still make sense even if we want to integration that into the debugging process later.

File: proto/pulumi/engine.proto [hunk 1]
```diff
@@ -83,3 +86,10 @@ message SetRootResourceRequest {
 message SetRootResourceResponse {
     // empty.
 }
+
+message StartDebuggerRequest {
+    // the debug configuration parameters.
```
Comment (Frassle) [inline proto/pulumi/engine.proto:91]: Should we make it clear these are DAP config values?

File: proto/pulumi/language.proto [hunk 2]
```diff
@@ -133,6 +133,7 @@ message RunRequest {
     string organization = 12;       // the organization of the stack being deployed into.
     google.protobuf.Struct configPropertyMap = 13; // the configuration variables to apply before running.
     ProgramInfo info = 14; // the program info to use to execute the program.
+    bool debug = 15;                // true if we're running in debug mode.
```
Comment (Frassle) [inline proto/pulumi/language.proto:136]: Feels like this should be more explicit like `start_debugger`

File: sdk/go/common/resource/plugin/context.go [hunk 3]
```diff
@@ -34,11 +34,12 @@ import (
...
 // appropriate. It also carries shared plugin configuration.
 type Context struct {
-	Diag       diag.Sink // the diagnostics sink to use for messages.
-	StatusDiag diag.Sink // the diagnostics sink to use for status messages.
-	Host       Host      // the host that can be used to fetch providers.
-	Pwd        string    // the working directory to spawn all plugins in.
-	Root       string    // the root directory of the context.
+	Diag       diag.Sink         // the diagnostics sink to use for messages.
+	StatusDiag diag.Sink         // the diagnostics sink to use for status messages.
+	Debugging  DebugEventEmitter // the debugging event emitter to use to signal the engine that the debugger started.
```
Comment (Frassle) [inline sdk/go/common/resource/plugin/context.go:39]: Shouldn't this only be needed on the Host, not the Context?

File: sdk/go/common/resource/plugin/debugging.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2016-2018, Pulumi Corporation.
```
Comment (julienp) [inline sdk/go/common/resource/plugin/debugging.go:1]:
```suggestion
// Copyright 2024, Pulumi Corporation.
```

**Comment by Frassle** [comment]
> but we don't have that available in the lifecycletests (? please let me know if I'm missing something)

Huh we don't, doesn't look simple to add unfortunately.

File: proto/pulumi/language.proto (hunk 2 above)
Comment (tgummerer) [reply proto/pulumi/language.proto:136]: Done :heavy_check_mark:

File: sdk/go/common/resource/plugin/context.go (hunk 3 above)
Comment (tgummerer) [reply sdk/go/common/resource/plugin/context.go]: Yeah, I think it was on the Context just for convenience, but it's not needed there.  Changed that in my last commit.

**Comment by tgummerer** [comment]
> Huh we don't, doesn't look simple to add unfortunately.

Might be fine to just implement some integration tests as we implement the other bits it in the language hosts.

File: proto/pulumi/language.proto [hunk 5]
```diff
@@ -132,7 +132,8 @@ message RunRequest {
     repeated string configSecretKeys = 11; // the configuration keys that have secret values.
     string organization = 12;       // the organization of the stack being deployed into.
     google.protobuf.Struct configPropertyMap = 13; // the configuration variables to apply before running.
-    ProgramInfo info = 14; // the program info to use to execute the program.
+    ProgramInfo info = 14;          // the program info to use to execute the program.
+    bool startDebugger = 15;       // true if the language host is supposed to start the program under a debugger.
```
[the remark this answers is not in the document]
Comment (tgummerer) [reply proto/pulumi/language.proto]: Ah I just followed what the other fields in this message do.  Fixed now :heavy_check_mark:

**Comment by tgummerer** [comment]
Had to rebase this to solve some conflicts.

**Review by Frassle — APPROVED** [review]

File: sdk/go/common/resource/plugin/host.go [hunk 6]
```diff
@@ -89,13 +89,17 @@ type Host interface {
 	// called before (e.g.) hard-closing any gRPC connection.
 	SignalCancellation() error
 
+	// StartDebugging asks the host to start a debugging session with the given configuration.
+	StartDebugging(DebuggingInfo) error
```
Comment (blampe) [inline sdk/go/common/resource/plugin/host.go:93]: Adding a new method to an interface like this is a breaking change, so upgrades are going to take longer than normal as we now need to manually implement this in a lot of places.

In the future I suggest implementing this as a standalone interface, and then test the host to see whether it supports debugging when you need it.

```go
type Host interface { /* unchanged */ }

type Debugger interface {
    StartDebugging(DebuggingInfo) error
}

...

func (eng *hostServer) StartDebugger(ctx context.Context,
 	req *pulumirpc.StartDebuggerRequest,
 ) (*emptypb.Empty, error) {
    debugger, ok := eng.host.(Debugger)
    if !ok {
        return &emptypb.Empty{}, nil
    }

 	err := debugger.StartDebugging(info)
 	if err != nil {
 		return nil, err
 	}
    return &emptypb.Empty{}, nil
}
```

**Pull request merged**